cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -329,6 +336,18 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
 
       emittedRecords.put(stageName, emittedBuilder.build());
     }
+
+    Collection<Future> sinkFutures = new ArrayList<>(sinkRunnables.size());
+    ExecutorService executorService = Executors.newFixedThreadPool(sinkRunnables.size(), new ThreadFactoryBuilder()
+      .setNameFormat("pipeline-sink-task")
+      .build());
+    for (Runnable runnable : sinkRunnables) {
+      sinkFutures.add(executorService.submit(runnable));
+    }
+    for (Future future : sinkFutures) {
+      future.get();
```

Comment:
If the task of this future failed, the `get()` call would throw, right? Meaning other tasks might still be running and there is no cancellation of them, and neither the executor service is being shutdown. I think it's better to have a `try-catch` around the `future.get()` and accumulate the exceptions and throw it out after all tasks are completed. Or if there is way to cancel the unfinished tasks, we should cancel them before throwing exception.